### PR TITLE
[core] LTX Video 0.9.1

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
To run conversion:

```
 python3 scripts/convert_ltx_to_diffusers.py --transformer_ckpt_path /raid/aryan/ltx-new/ltx-video-2b-v0.9.1.safetensors --vae_ckpt_path /raid/aryan/ltx-new/ltx-video-2b-v0.9.1.safetensors --output_path /raid/aryan/ltx-diffusers --dtype bf16 --version 0.9.1 --text_encoder_cache_dir /raid/.cache/huggingface/ --save_pipeline
```

(I've verified that the conversion for v0.9.0 still works after the current modifications to the script)

Inference after conversion:

```python
import torch
from diffusers import LTXPipeline
from diffusers.utils import export_to_video

pipe = LTXPipeline.from_pretrained("/raid/aryan/ltx-diffusers", torch_dtype=torch.bfloat16)
pipe.to("cuda")

prompt = "A woman with long brown hair and light skin smiles at another woman with long blonde hair. The woman with brown hair wears a black jacket and has a small, barely noticeable mole on her right cheek. The camera angle is a close-up, focused on the woman with brown hair's face. The lighting is warm and natural, likely from the setting sun, casting a soft glow on the scene. The scene appears to be real-life footage"
negative_prompt = "worst quality, inconsistent motion, blurry, jittery, distorted"

video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    width=704,
    height=480,
    num_frames=161,
    num_inference_steps=50,
    decode_timestep=0.05,
    generator=torch.Generator(device="cuda").manual_seed(0),
).frames[0]
export_to_video(video, "output.mp4", fps=24)
```

Output on prompts from the model page:

<table align=center>
<td>
<video src="https://github.com/user-attachments/assets/6610eb45-c84a-4b8c-b6ed-740958e1f2fa"> Your browser does not support the video tag.</video>
</td>
</table>

Will open weights PR to the [official repository](https://huggingface.co/Lightricks/LTX-Video) soon.

cc @yoavhacohen @SapirW

### CODE REVIEW

File: src/diffusers/models/autoencoders/autoencoder_kl_ltx.py [hunk 1]
```diff
@@ -433,15 +540,32 @@ def __init__(
...
+        hidden_states: torch.Tensor,
+        temb: Optional[torch.Tensor] = None,
+        generator: Optional[torch.Generator] = None,
+    ) -> torch.Tensor:
         if self.conv_in is not None:
-            hidden_states = self.conv_in(hidden_states)
+            hidden_states = self.conv_in(hidden_states, temb, generator)
+
+        if self.time_embedder is not None:
+            temb = self.time_embedder(
```
Comment (yiyixuxu) [inline src/diffusers/models/autoencoders/autoencoder_kl_ltx.py:561]: ohhh, they run this on block level? 🤔

File: src/diffusers/pipelines/ltx/pipeline_ltx.py [hunk 2]
```diff
@@ -563,6 +565,10 @@ def __call__(
                 provided, negative_prompt_embeds will be generated from `negative_prompt` input argument.
             negative_prompt_attention_mask (`torch.FloatTensor`, *optional*):
                 Pre-generated attention mask for negative text embeddings.
+            decode_timestep (`float`, defaults to `0.05`):
+                The timestep at which generated video is decoded.
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx/pipeline_ltx.py:569]: what does this mean?

File: src/diffusers/pipelines/ltx/pipeline_ltx.py [hunk 3]
```diff
@@ -753,7 +759,25 @@ def __call__(
...
             latents = latents.to(prompt_embeds.dtype)
-            video = self.vae.decode(latents, return_dict=False)[0]
+
+            if not self.vae.config.timestep_conditioning:
+                timestep = None
+            else:
+                noise = torch.randn(latents.shape, generator=generator, device=device, dtype=latents.dtype)
+                if not isinstance(decode_timestep, list):
+                    decode_timestep = [decode_timestep] * batch_size
+                if decode_noise_scale is None:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx/pipeline_ltx.py:769]: it is strange for `None` to have a special behavior like this. can we default it to `None`?

**Review by yiyixuxu — COMMENTED** [review]
that's fast! thanks!

**Comment by tin2tin** [comment]
Are the 0.9.1 diffusers weights up on HuggingFace?

File: src/diffusers/models/autoencoders/autoencoder_kl_ltx.py (hunk 1 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/autoencoders/autoencoder_kl_ltx.py:561]: Yep, it's in the decoder as well as the mid and up blocks

File: src/diffusers/pipelines/ltx/pipeline_ltx.py (hunk 3 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/ltx/pipeline_ltx.py:769]: Yes that sounds good. I'll make the example use the recommended default value of 0.025 though as demo

**Comment by a-r-r-o-w** [comment]
@tin2tin We're still working with their team on how to host the weights. It might take some time :(

Until then, the model is available here as well: https://huggingface.co/a-r-r-o-w/LTX-Video-0.9.1-diffusers, mostly because I need this weight format for finetuning. Once we have it hosted officially, those can be used instead

**Comment by nitinmukesh** [comment]
> Until then, the model is available here as well: https://huggingface.co/a-r-r-o-w/LTX-Video-0.9.1-diffusers, mostly because I need this weight format for finetuning. Once we have it hosted officially, those can be used instead

Thank you for sharing.

**Comment by a-r-r-o-w** [comment]
cc @DN6 for single-file related support

**Review by yiyixuxu — APPROVED** [review]

**Comment by tin2tin** [comment]
Wow, this 0.9.1 (with prompt input) delivers very good quality video, and very fast!

I guess img2vid is not implemented yet - I'm getting this error - sorry, if this is just premature to test this:
```
Python311\site-packages\diffusers\models\autoencoders\autoencoder_kl_ltx.py", line 431, in forward
    timestep=temb.flatten(),
             ^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'flatten'
```

**Comment by a-r-r-o-w** [comment]
Oh, taking a look... This PR should worked for I2V as well

**Comment by a-r-r-o-w** [comment]
@tin2tin Could you try again? Thanks for catching!

**Review by DN6 — APPROVED** [review]

**Comment by tin2tin** [comment]
@a-r-r-o-w That was quick! Yes, it is fixed now! Thank you!

https://github.com/user-attachments/assets/35e96e97-6b2b-425b-809a-d409ffbfe82d

**Comment by scarbain** [comment]
> @tin2tin We're still working with their team on how to host the weights. It might take some time :(
> 
> Until then, the model is available here as well: https://huggingface.co/a-r-r-o-w/LTX-Video-0.9.1-diffusers, mostly because I need this weight format for finetuning. Once we have it hosted officially, those can be used instead

Hi @a-r-r-o-w, thanks for this PR merged ! Any script for finetuning or LoRA i2v available somewhere ? :)

**Comment by a-r-r-o-w** [comment]
@scarbain There's one for T2V here: https://github.com/a-r-r-o-w/finetrainers. I2V will be supported soon!

**Pull request merged**